cocotb/share/lib/verilator/verilator.cpp

```diff
@@ -57,43 +75,66 @@ int main(int argc, char** argv) {
 #endif
 
     while (!Verilated::gotFinish()) {
-        bool again = true;
+        // Call registered timed callbacks (e.g. clock timer)
+        // These are called at the beginning of the time step
+        // before the iterative regions (IEEE 1800-2012 4.4.1)
+        VerilatedVpi::callTimedCbs();
+
+        // Call Value Change callbacks triggered by Timer callbacks
+        // These can modify signal values
+        settle_value_callbacks();
 
         // We must evaluate whole design until we process all 'events'
+        bool again = true;
         while (again) {
             // Evaluate design
             top->eval();
 
-            // Call Value Change callbacks as eval()
-            // can modify signals values
-            VerilatedVpi::callValueCbs();
+            // Call Value Change callbacks triggered by eval()
+            // These can modify signal values
+            again = settle_value_callbacks();
 
-            // Call registered Read-Write callbacks
-            again = VerilatedVpi::callCbs(cbReadWriteSynch);
+            // Call registered ReadWrite callbacks
+            again |= VerilatedVpi::callCbs(cbReadWriteSynch);
 
-            // Call Value Change callbacks as cbReadWriteSynch
-            // can modify signals values
-            VerilatedVpi::callValueCbs();
+            // Call Value Change callbacks triggered by ReadWrite callbacks
+            // These can modify signal values
+            again |= settle_value_callbacks();
         }
 
         // Call ReadOnly callbacks
         VerilatedVpi::callCbs(cbReadOnlySynch);
 
-        // Call registered timed callbacks (e.g. clock timer)
-        VerilatedVpi::callTimedCbs();
-
 #if VM_TRACE
         tfp->dump(main_time);
 #endif
-        main_time++;
+        // cocotb controls the clock inputs using cbAfterDelay so
+        // skip ahead to the next registered callback
+        vluint64_t next_time = VerilatedVpi::cbNextDeadline();
+
+        // If there are no more cbAfterDelay callbacks,
+        // the next deadline is max value, so end the simulation now
+        if (next_time == static_cast<vluint64_t>(~0ULL)) {
+            Verilated::gotFinish(true);
```

Comment:
Does Verilator not support `vpi_control(vpiFinish, 0)`? Is that not desirable here? Are `cbEndOfSimulation` callbacks called in this case?

Reply:
`vl_finish()` may be more appropriate, which `vpi_control(vpiFinish, 0)` ends up calling. And we are breaking the loop here anyway.

`vpi_control()` is for controlling the simulator from a VPI application, but this context is the simulator ending the simulation because there isn't any more stimulus.

`vl_finish()` also flushes stdout/stderr and VCD/FST files, so I think that's the better option.

Either way, `cbEndOfSimulation` callbacks are explicitly called below on line 134.